datasquare.go

```diff
@@ -173,13 +181,38 @@ func (ds *dataSquare) resetRoots() {
 }
 
 func (ds *dataSquare) computeRoots() {
+	var wg sync.WaitGroup
+	var mu sync.Mutex
+
 	rowRoots := make([][]byte, ds.width)
 	colRoots := make([][]byte, ds.width)
+
 	for i := uint(0); i < ds.width; i++ {
-		rowRoots[i] = ds.getRowRoot(i)
-		colRoots[i] = ds.getColRoot(i)
+		i := i
+		wg.Add(2)
+
+		go func() {
+			defer wg.Done()
+
+			rowRoot := ds.getRowRoot(i)
+
+			mu.Lock()
+			defer mu.Unlock()
+			rowRoots[i] = rowRoot
```

Comment:
You don't need to lock the slice here because each index writes to its memory slot of the slice. This way removes any contention between routines, what really matters for high throughput

Reply:
Can find are few links to prove this, as no need for mutexes can sound counterintuitive at the first sight here